cluster-autoscaler/core/loop_state.go

```diff
@@ -0,0 +1,448 @@
+	potentiallyUnneeded := getPotentiallyUnneededNodes(ls.context, ls.allNodes)
+
+	typedErr := ls.autoscaler.scaleDown.UpdateUnneededNodes(
+		ls.allNodes, potentiallyUnneeded,
+		append(ls.allScheduled, ls.unschedulableWaitingForLowerPriorityPreemption...),
+		ls.currentTime, pdbs)
+	if typedErr != nil {
+		glog.Errorf("Failed to scale down: %v", typedErr)
+		ls.onTypedError(typedErr)
+	}
+
+	metrics.UpdateDurationFromStart(metrics.FindUnneeded, unneededStart)
+
+	if glog.V(4) {
+		for key, val := range ls.autoscaler.scaleDown.unneededNodes {
+			glog.V(4).Infof("%s is unneeded since %s duration %s", key, val.String(), ls.currentTime.Sub(val).String())
+		}
+	}
+
+	// In dry run only utilization is updated
+	calculateUnneededOnly := ls.scaleDownForbidden ||
+		ls.autoscaler.lastScaleUpTime.Add(ls.autoscaler.ScaleDownDelayAfterAdd).After(ls.currentTime) ||
+		ls.autoscaler.lastScaleDownFailTime.Add(ls.autoscaler.ScaleDownDelayAfterFailure).After(ls.currentTime) ||
+		ls.autoscaler.lastScaleDownDeleteTime.Add(ls.autoscaler.ScaleDownDelayAfterDelete).After(ls.currentTime) ||
+		ls.autoscaler.scaleDown.nodeDeleteStatus.IsDeleteInProgress()
+
+	glog.V(4).Infof("Scale down status: unneededOnly=%v lastScaleUpTime=%s "+
+		"lastScaleDownDeleteTime=%v lastScaleDownFailTime=%s scaleDownForbidden=%v isDeleteInProgress=%v",
+		calculateUnneededOnly, ls.autoscaler.lastScaleUpTime, ls.autoscaler.lastScaleDownDeleteTime, ls.autoscaler.lastScaleDownFailTime,
+		ls.scaleDownForbidden, ls.autoscaler.scaleDown.nodeDeleteStatus.IsDeleteInProgress())
+
+	if calculateUnneededOnly {
+		return ls.onSuccess()
+	}
+
+	glog.V(4).Infof("Starting scale down")
+
+	// We want to delete unneeded Node Groups only if there was no recent scale up,
+	// and there is no current delete in progress and there was no recent errors.
+	if ls.context.NodeAutoprovisioningEnabled {
+		err := cleanUpNodeAutoprovisionedGroups(ls.context.CloudProvider, ls.context.LogRecorder)
+		if err != nil {
+			glog.Warningf("Failed to clean up unneeded node groups: %v", err)
+		}
+	}
+
+	scaleDownStart := time.Now()
+	metrics.UpdateLastTime(metrics.ScaleDown, scaleDownStart)
+	result, typedErr := ls.autoscaler.scaleDown.TryToScaleDown(ls.allNodes, ls.allScheduled, pdbs, ls.currentTime)
+	metrics.UpdateDurationFromStart(metrics.ScaleDown, scaleDownStart)
+
+	// TODO: revisit result handling
+	if typedErr != nil {
+		glog.Errorf("Failed to scale down: %v", err)
+		ls.autoscaler.lastScaleDownFailTime = ls.currentTime
+		return ls.onTypedError(typedErr)
+	}
+	if result == ScaleDownNodeDeleted || result == ScaleDownNodeDeleteStarted {
+		ls.autoscaler.lastScaleDownDeleteTime = ls.currentTime
+	}
+
+	return ls.onSuccess()
+}
+
+// single-use helper methods start here, extracted to limit complexity of any single method
+// all these methods return the values used by phase methods of LoopStateHandler interface
+
+func (ls *loopState) refreshCloudProvider() (ExecutionResult, errors.AutoscalerError) {
```

Comment:
Why does refreshCloudProvider return ExecutionResult enum? It can't break out of main loop without error

Reply:
It's just to have a consistent interface - I'd rather avoid exposing implementation detail (which might only temporarily hold) to the outside.